src/multiplex/multiplex.rs

```diff
@@ -377,75 +451,175 @@ impl<S, T, E> Multiplex<S, T>
     fn write_in_messages(&mut self) -> io::Result<()> {
         trace!("write in messages");
 
-        while self.transport.poll_write().is_ready() {
+        while self.dispatch.transport().poll_write().is_ready() {
             trace!("   --> polling for in frame");
 
-            if let Some((id, msg)) = self.dispatch.poll() {
-                try!(self.write_in_message(id, msg));
-            } else {
-                break;
+            match try!(self.dispatch.poll()) {
+                Async::Ready(Some((id, Ok(message)))) => {
+                    self.dispatch_made_progress();
+
+                    trace!("   --> got message");
+                    try!(self.write_in_message(id, message));
+                }
+                Async::Ready(Some((id, Err(error)))) => {
+                    self.dispatch_made_progress();
+
+                    trace!("   --> got error");
+                    try!(self.write_in_error(id, error));
+                }
+                Async::Ready(None) => {
+                    trace!("   --> got error");
+                    trace!("   --> got None");
+                    // The service is done with the connection. In this case, a
+                    // `Done` frame should be written to the transport and the
+                    // transport should start shutting down.
+                    //
+                    // However, the `Done` frame should only be written once
+                    // all the in-flight bodies have been written.
+                    //
+                    // For now, do nothing...
```

Comment:
Could you tag this with a TODO? I imagine this'd be something like set a flag and then check it after all bodies are written out?